### DISCUSSION TITLE
[@types/react] Introducing tag for jsdocs for relevant attributes to a html element

### DISCUSSION THREAD

**Discussion by 0xRezaa** [DefinitelyTyped/DefinitelyTyped#65452] [Issues with a @types package]
**Context**

We at [Codux](https://www.codux.com/) encountered the need to differentiate between html attributes based on there relevance for a html element. Question at hand is something like: Which attributes of a `div` element are more relevant than the others? The same question can be asked about `button`, `img`, etc.

The same question can also be asked in general: Are there HTML attributes which are more relevant then others in every case.
For example: `className` or `style` attribute.

We tried to tackle the problem via adding information about the attributes in the jsdocs of each attribute. Because this information could be considered specific to our application we wanted to move it into it's own type package `@types/codux-react` and add the jsdocs through augmentation. See this [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65271). After some time it seemed to us that it would be beneficial to have an approach which the whole community can decide on and benefit from. 

**Suggestion**

Our suggestion is to agree on a [jsdocs block tag](https://jsdoc.app/about-block-inline-tags.html) definition to describe the relevance of an attribute agnostic of our application.

We would then have an `RelevantTag` followed by possibly more information. For example a list of HTML elements the attribute is relevant for. This could look somewhat like:

```
/**
 * @relevant div, span, button
 */
attribute: AttributeType;
```

Other possible candidates for the `RelevantTag` could be: important, relevantfor, importantfor.

**Community**

What do you people think about this approach. Do you have any suggestions or concerns?

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
What would make an attribute relevant for some tags but not for others? 

Could you include some concrete examples for problems you're trying to solve with this proposal?

**Reply by 0xRezaa** [reply]
To make a case on attributes being relevant for some html elements but not for others I would look at three cases:
1. Some attributes are exclusive for certain elements. Like the `href` attribute which is only present on following interfaces: `LinkHTMLAttributes`, `SVGAttributes`, `BaseHTMLAttributes`, `AreaHTMLAttributes` and `AnchorHTMLAttributes`. For a common developer/user there is no way to get this relationship without looking into the documentation.
2. Some attributes are commonly used with certain html elements like `href` with an Anchor element or `src` with an Img element.
3. Some attributes are used more often in day to day life. For example the `className` attribute is more likely to be used on a day to day basis on any element over for example `draggable`. This is of course opinionated but I think the community could agree on a heuristic for the relevance for attributes for certain elements.

Regarding the problems we are trying to solve:
We are working on a visual React IDE and in one part of it we display properties of an selected element/component out of the tree of rendered elements/components. To provide the user with the most relevant properties we are parsing the types of these and understand from where they are coming from. We would like to also parse this "meta information" about relevance from the jsdocs of the attributes.

As a concrete example:
If the user has selected an `<img />` element we would like to show the `src` attribute over others.

**Reply by eps1lon** [reply]
> Some attributes are exclusive for certain elements.

We should only declare them for the elements that only allow them exclusively. I don't understand how special JSDoc would help here.

> Some attributes are commonly used with certain html elements like href with an Anchor element or src with an Img element.

Sure but this "commonly" definition needs to be fleshed out more.

> Some attributes are used more often in day to day life. For example the className attribute is more likely to be used on a day to day basis on any element over for example draggable.

Yeah I definitely get that. But it isn't clear to me how some new kind of JSDoc would help here. It seems to me that you want to open this issue against TypeScript or VSCode instead so that we can come to an ecosystem wide understanding. Doing something exclusive to `@types/react` and your IDE will just add maintenance burden for us without having a community-wide understanding of what these special JSDoc annotations are for.
